File: sdk/plugin/source.go
package plugin

// Wire contract for CapabilitySource plugins. The host bridges a `type:
// plugin` source in apiary.yaml to one plugin instance and invokes these
// methods over the standard single-shot protocol; the plugin polls its
// backing system and returns work items. Source plugins are read-only, like
// the in-tree monitoring sources: items they surface cannot host approval
// steps, CI waits, or label/state write-back.
const (
	// SourceMethodPoll returns the current work items (SourcePollResult).
	// Invoked on the source's poll interval with a SourcePollRequest. Items
	// must carry stable IDs: the host dispatches each distinct ID at most
	// once per its dedup rules, and an item returned on every poll while it
	// stays relevant is the expected shape (mirror of the in-tree
	// monitoring adapters).
	SourceMethodPoll = "poll"

	// SourceMethodAcknowledge is invoked after an item has been dispatched
	// (SourceAckRequest). Plugins with nothing to mark should return
	// SourceOKResult{OK: true} rather than an error.
	SourceMethodAcknowledge = "acknowledge"

	// SourceMethodWriteResult is invoked with the run outcome for an item
	// (SourceWriteResultRequest). Plugins with no result surface should
	// return SourceOKResult{OK: true} rather than an error.
	SourceMethodWriteResult = "write_result"
)

// SourceItem is one unit of work surfaced by a source plugin.
type SourceItem struct {
	// ID is the source-native identifier and the dedup key — it must be
	// stable for the lifetime of the item and unique per dispatch-worthy
	// occurrence. Items without an ID are dropped by the host.
	ID string `json:"id"`
	// Number is the human-facing reference (e.g. "INC-42"). Defaults to ID.
	Number string `json:"number,omitempty"`
	Title  string `json:"title,omitempty"`
	// Description becomes the task body handed to the agent.
	Description string `json:"description,omitempty"`
	// Labels drive workflow trigger matching; use "key:value" form.
	Labels   []string `json:"labels,omitempty"`
	Type     string   `json:"type,omitempty"`
	Priority string   `json:"priority,omitempty"`
	State    string   `json:"state,omitempty"`
	URL      string   `json:"url,omitempty"`
	// Metadata is carried opaquely onto the task.
	Metadata map[string]any `json:"metadata,omitempty"`
	// CreatedAt / UpdatedAt are RFC3339 timestamps; empty means "now".
	CreatedAt string `json:"created_at,omitempty"`
	UpdatedAt string `json:"updated_at,omitempty"`
}

// SourcePollRequest is the payload of SourceMethodPoll.
type SourcePollRequest struct {
	// Since is the previous poll time (RFC3339), empty on the first poll.
	// Informational: plugins that re-return ongoing items may ignore it.
	Since string `json:"since,omitempty"`
	// States / Labels are the source's `filters` from apiary.yaml, forwarded
	// so the plugin can filter at the backend where possible.
	States []string `json:"states,omitempty"`
	Labels []string `json:"labels,omitempty"`
}

// SourcePollResult is the result of SourceMethodPoll.
type SourcePollResult struct {
	Items []SourceItem `json:"items"`
}

// SourceAckRequest is the payload of SourceMethodAcknowledge.
type SourceAckRequest struct {
	Item SourceItem `json:"item"`
	// Action is the host's ack action (e.g. "dispatched").
	Action string `json:"action"`
}

// SourceWriteResultRequest is the payload of SourceMethodWriteResult.
type SourceWriteResultRequest struct {
	Item    SourceItem `json:"item"`
	Success bool       `json:"success"`
	Output  string     `json:"output,omitempty"`
	Error   string     `json:"error,omitempty"`
}

// SourceOKResult is the conventional result for acknowledge/write_result.
type SourceOKResult struct {
	OK bool `json:"ok"`
}

File: sdk/python/apiary_plugin/source.py
"""Typed mirrors of the ``source`` capability wire contract.

A line-for-line counterpart of ``sdk/plugin/source.go``. Optional fields are
omitted from the encoded JSON, matching the Go structs' ``omitempty`` tags, so
a plugin written against either SDK puts the same bytes on the wire.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, List, Mapping, Optional

# Method names a source plugin switches on.
SOURCE_METHOD_POLL = "poll"
SOURCE_METHOD_ACKNOWLEDGE = "acknowledge"
SOURCE_METHOD_WRITE_RESULT = "write_result"


def _compact(values: dict) -> dict:
    """Drop empty optional fields — the `omitempty` of the Go structs."""
    return {key: value for key, value in values.items() if value not in (None, "", [], {})}


@dataclass
class SourceItem:
    """One unit of work surfaced by a source plugin.

    ``id`` is the source-native identifier **and the host's dedup key**: it must
    be stable for the lifetime of the item and unique per dispatch-worthy
    occurrence. Items without one are dropped by the host, so it is the only
    required field. Returning the same item on every poll while it stays
    relevant is the expected shape.

    ``labels`` drive workflow trigger matching and use ``key:value`` form.
    ``created_at`` / ``updated_at`` are RFC3339 strings; empty means "now".
    """

    id: str
    number: str = ""
    title: str = ""
    description: str = ""
    labels: List[str] = field(default_factory=list)
    type: str = ""
    priority: str = ""
    state: str = ""
    url: str = ""
    metadata: Mapping[str, Any] = field(default_factory=dict)
    created_at: str = ""
    updated_at: str = ""

    def to_dict(self) -> dict:
        encoded = _compact(
            {
                "number": self.number,
                "title": self.title,
                "description": self.description,
                "labels": list(self.labels),
                "type": self.type,
                "priority": self.priority,
                "state": self.state,
                "url": self.url,
                "metadata": dict(self.metadata),
                "created_at": self.created_at,
                "updated_at": self.updated_at,
            }
        )
        # id is never omitted: an item without one is not dispatchable.
        return {"id": self.id, **encoded}

    @classmethod
    def from_dict(cls, raw: Mapping[str, Any]) -> "SourceItem":
        return cls(
            id=raw.get("id", ""),
            number=raw.get("number", ""),
            title=raw.get("title", ""),
            description=raw.get("description", ""),
            labels=list(raw.get("labels") or []),
            type=raw.get("type", ""),
            priority=raw.get("priority", ""),
            state=raw.get("state", ""),
            url=raw.get("url", ""),
            metadata=dict(raw.get("metadata") or {}),
            created_at=raw.get("created_at", ""),
            updated_at=raw.get("updated_at", ""),
        )


@dataclass
class SourcePollRequest:
    """Payload of ``poll``.

    ``since`` is the previous poll time (RFC3339, empty on the first poll) and
    is informational: plugins that re-return ongoing items may ignore it.
    ``states`` / ``labels`` are the source's ``filters:`` from apiary.yaml,
    forwarded so the plugin can filter at the backend where possible.
    """

    since: str = ""
    states: List[str] = field(default_factory=list)
    labels: List[str] = field(default_factory=list)

    @classmethod
    def from_dict(cls, raw: Optional[Mapping[str, Any]]) -> "SourcePollRequest":
        raw = raw or {}
        return cls(
            since=raw.get("since", ""),
            states=list(raw.get("states") or []),
            labels=list(raw.get("labels") or []),
        )


@dataclass
class SourcePollResult:
    """Result of ``poll``: the plugin's current work items."""

    items: List[SourceItem] = field(default_factory=list)

    def to_dict(self) -> dict:
        # `items` is always present, empty array included — "no work right now"
        # is an answer, and dropping the key would look like a malformed result.
        return {"items": [item.to_dict() for item in self.items]}


@dataclass
class SourceAckRequest:
    """Payload of ``acknowledge``: the item and the host's ack action."""

    item: SourceItem
    action: str = ""

    @classmethod
    def from_dict(cls, raw: Optional[Mapping[str, Any]]) -> "SourceAckRequest":
        raw = raw or {}
        return cls(item=SourceItem.from_dict(raw.get("item") or {}), action=raw.get("action", ""))


@dataclass
class SourceWriteResultRequest:
    """Payload of ``write_result``: the run outcome for an item."""

    item: SourceItem
    success: bool = False
    output: str = ""
    error: str = ""

    @classmethod
    def from_dict(cls, raw: Optional[Mapping[str, Any]]) -> "SourceWriteResultRequest":
        raw = raw or {}
        return cls(
            item=SourceItem.from_dict(raw.get("item") or {}),
            success=bool(raw.get("success", False)),
            output=raw.get("output", ""),
            error=raw.get("error", ""),
        )


@dataclass
class SourceOKResult:
    """The conventional ``{"ok": true}`` answer to a write-back method."""

    ok: bool = True

    def to_dict(self) -> dict:
        return {"ok": self.ok}
